File: src/prismic.tests/ApiTest.fs
namespace prismic.tests
open prismic
open System
open NUnit.Framework

module ApiTest =

    let await a = a |> Async.RunSynchronously
    let expectException statement matchesException = 
        try
            statement() |> ignore
            Assert.Fail("expected exception was not raised")
        with | e -> match matchesException(e) with
                    | true -> Assert.That(true)
                    | _ -> reraise()//Assert.Fail(sprintf "unexpected type of exception happened: %s %s" (e.GetType().Name) e.Message)
    
    let apiGetNoCache = Api.get (ApiInfra.NoCache() :> ApiInfra.ICache<Api.Response>) (ApiInfra.Logger.NoLogger)

    [<TestFixture>]
    type ``Get Private Api``() = 

        [<Test>]
        member x.``Without Authorization Token Should Throw``() =
            let url = "https://private-test.prismic.io/api"
            expectException 
                (fun () -> await (apiGetNoCache (Option.None) url))
                (function
                    | Api.AuthorizationNeeded(_, url) -> url = "https://private-test.prismic.io/auth"
                    | e -> false)

        [<Test>]
        member x.``With Invalid Token Should Throw``() =
            let url = "https://private-test.prismic.io/api"
            expectException 
                (fun () -> await (apiGetNoCache (Option.Some("dummy-token")) url))
                (function
                    | Api.InvalidToken(_, url) -> url = "https://private-test.prismic.io/auth"
                    | e -> false)

File: src/prismic.csharp.tests/ApiTest.cs
using NUnit.Framework;
using NUnit.Framework.Constraints;
using System;
using System.Threading.Tasks;
using prismic.extensions;

namespace prismic.csharp.tests
{
	[TestFixture ()]
	public class ApiTest
	{
		[Test ()]
		public void GetPrivateApiWithoutAuthorizationTokenShouldThrow()
		{
			var url = "https://private-test.prismic.io/api";

			ExpectInnerException<Api.AuthorizationNeeded>( 
				() => prismic.extensions.Api.Get (url, new prismic.Infra.NoCache<prismic.Api.Response>(), (l, m) => {}).Wait(),
				e => 
				"https://private-test.prismic.io/auth" == e.Data0);
		}

		[Test ()]
		public void GetPrivateApiWithInvalidTokenShouldThrow()
		{
			var url = "https://private-test.prismic.io/api";

			ExpectInnerException<Api.InvalidToken>( 
				() => prismic.extensions.Api.Get ("dummy-token", url, new prismic.Infra.NoCache<prismic.Api.Response>(), (l, m) => {}).Wait(),
				e => 
				"https://private-test.prismic.io/auth" == e.Data0);
		}

		private void ExpectInnerException<ExT>(Action action, Func<ExT, bool> exceptionPredicate) where ExT : Exception
		{
			try {
				ThrowInner(action);
				Assert.Fail("expected exception was not raised");
			} catch (ExT ex) {
				exceptionPredicate (ex);
			} catch (Exception ex) {
				Assert.Fail(String.Format("unexpected type of exception happened: {0} {1}", ex.GetType().Name, ex.Message));
			}
		}

		private void ThrowInner(Action action) 
		{
			try {
				action();
			} catch (AggregateException ex) {
				throw ex.InnerException;
			}
		}

	}
}

